medic/medic-webapp
# #4445 - Reduce time to render contacts list through live-list

# Description

#4445

Performance optimizations for the Live-List class as-used in the contacts tab. 

Contact controller currently calls `livelist.update` once per list item and then `livelist.refresh` after all items are added. 

This change updates the interface `livelist.set` to avoid the incremental dom changes from `livelist.update` and then the full re-paint from `livelist.refresh`. The proposed change inserts the items each item into the model once, sorts the full list once, and then draws everything once. To avoid unnecessary scanning, update `livelist.dom` to a hashmap. `livelist.set()` accepts an optional flag `reuseExistingDom` which can avoid expensive calls to `listItemFor` when the contents of the DOM are being reused and not updated (eg. pagination). 

## Measurements
On my SM-G361H. Times are a measure of milliseconds spent in `live-list.js` functions.

Five samples of the time spent in Live-List drawing a full LHS contact list (40 contacts).

Sample | Proposed | Master
-- | -- | --
1 | 391 | 546
2 | 304 | 660
3 | 261 | 856
4 | 277 | 499
5 | 387 | 800
Mean | 324 | 660

Two hundred contacts, displayed over four paginated pages. Table shows the time to render each page in the LHS contact list while paginating.

Page | Proposed | Master
-- | -- | --
1 | 519 | 725
2 | 419 | 738
3 | 464 | 876
4 | 311 | 671
Mean | 428 | 752

## Testing
I'm testing the following scenarios:

* Create a new user as admin. Syncs to offline user's live-list. 
* Delete the user as admin. Syncs to offline user's live-list.
* Edit user as admin. Syncs to offline user's live-list.
* Edit user twice as offline user. Edits appear in live-list, no pagination.
* Proper pagination.
* Edit the home place as admin (doesn't update in offline user's live-list) 
* Paginate to nth page, then edit an element within that page.
* Search
* Update the name of a primary contact for clinic

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.
